### ISSUE TITLE
Type Narrowing Failure

### ISSUE THREAD

**Issue by SirRegion** [microsoft/pyright#9593]
**Describe the bug**
Pyright does not narrow the type of the returned value correctly.

**Code**

```python
def simple_test(value: str | None, boolean: bool) -> str:
   
    if boolean is False:
       return "false"
    
    if boolean is True and value is None:
        return "true"
    
    return value
```
In this case, the return value can only ever be a string. However, pyright is not able to correctly deduce that and fails with ` Type "str | None" is not assignable to return type "str"`

You can see all possible values in the table below:
| value  | boolean | returned |
|--------|---------|---------|
| None   | False   | "false" |
| None   | True    | "true"  |
| "test" | False   | "false" |
| "test" | True    | "test"  |

**VS Code extension or command-line**
This fails in PyCharm via the PyCharm plugin and on the command line.

**Workaround**

This issue can be worked around by adding an additional, if unneccessary, check to the function and an unreachable return:

```python
def simple_test(value: str | None, boolean: bool) -> str:
   
    if boolean is False:
       return "false"
    
    if boolean is True and value is None:
        return "true"
    if value is not None:
        return value
    return "unreachable"
```
or by ignoring the return with `# pyright: ignore [reportReturnType]`.

**Comment by erictraut** [comment]
Pyright is working as designed here, so this isn't a bug.

Static type checkers like pyright do not do "mutual type narrowing" (what I sometimes refer to as "entanglement"). That is, a variable's narrowed type doesn't depend on the narrowed type of another variable. Expressions are narrowed independent of each other. While it would be theoretically possible to track "dependent types" across symbols, it would be computationally infeasible. I'm not aware of any type checker or compiler in any language that does this.

As you've pointed out, there are ways to write your code to avoid depending on "entanglement" between two variables.

**Comment by SirRegion** [comment]
Thank you for your quick response, I did not know that. Do you have any resources I could read to further my knowledge? 


Btw:
Suprisingly, MyPy is fine with the above code snipped (at least in the online checker I tried it with).
